File: binary-tree/binary-tree-right-side-view.py
# Definition for a binary tree node.
# class TreeNode:
#     def __init__(self, val=0, left=None, right=None):
#         self.val = val
#         self.left = left
#         self.right = right
class Solution:
    def rightSideView(self, root: Optional[TreeNode]) -> List[int]:
        if root is None:
            return []
        visited = set()
        q = collections.deque()
        q.append((root,0))
        ret = [root.val]
        self.check_height = 1 
        while q:
            cur,height = q.popleft()
            if cur.right is not None: 
                q.append((cur.right,height + 1)) 
                if self.check_height == height + 1:
                    ret.append(cur.right.val) 
                    self.check_height+=1 
            if cur.left is not None: 
                q.append((cur.left, height + 1)) 
                if self.check_height == height + 1:
                    ret.append(cur.left.val) 
                    self.check_height+=1 
        return ret
                
        

                
            

            
        return ret


        

File: breadth-first-search/binary-tree-right-side-view.java
/**
 * Definition for a binary tree node.
 * public class TreeNode {
 *     int val;
 *     TreeNode left;
 *     TreeNode right;
 *     TreeNode() {}
 *     TreeNode(int val) { this.val = val; }
 *     TreeNode(int val, TreeNode left, TreeNode right) {
 *         this.val = val;
 *         this.left = left;
 *         this.right = right;
 *     }
 * }
 */
class Solution {
    public List<Integer> rightSideView(TreeNode root) {
        if (root == null){
            return new ArrayList<>(); 
        }
        List<Integer> res = new ArrayList<>(); 
        Queue<TreeNode> queue = new ArrayDeque<>(); 
        queue.add(root); 
        
        while(!queue.isEmpty()){
            int queueSize = queue.size(); 
            for(int i = 0; i < queueSize; i++){
                TreeNode cur = queue.poll(); 
                if (i == queueSize-1){
                    res.add(cur.val); 
                }
                if (cur.left != null){
                    queue.add(cur.left); 
                }
                if (cur.right != null){
                    queue.add(cur.right); 
                }
            }
        }
        return res; 
    }
}